Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift

```diff
@@ -0,0 +1,634 @@
+            return self.createNewConnection()
+        }
+
+        @inlinable
+        mutating func createNewOverflowConnectionIfPossible() -> ConnectionRequest? {
+            precondition(self.maximumConcurrentConnectionSoftLimit <= self.stats.active)
+            guard self.maximumConcurrentConnectionHardLimit > self.stats.active else {
+                return nil
+            }
+            return self.createNewConnection()
+        }
+
+        @inlinable
+        /*private*/ mutating func createNewConnection() -> ConnectionRequest {
+            precondition(self.canGrow)
+            self.stats.connecting += 1
+            let connectionID = self.generator.next()
+            let connection = ConnectionState(id: connectionID)
+            self.connections.append(connection)
+            return ConnectionRequest(connectionID: connectionID)
+        }
+
+        /// A new ``Connection`` was established.
+        ///
+        /// This will put the connection into the idle state.
+        ///
+        /// - Parameter connection: The new established connection.
+        /// - Returns: An index and an IdleConnectionContext to determine the next action for the now idle connection.
+        ///            Call ``parkConnection(at:)``, ``leaseConnection(at:)`` or ``closeConnection(at:)``
+        ///            with the supplied index after this.
+        @inlinable
+        mutating func newConnectionEstablished(_ connection: Connection, maxStreams: UInt16) -> (Int, AvailableConnectionContext) {
+            guard let index = self.connections.firstIndex(where: { $0.id == connection.id }) else {
+                preconditionFailure("There is a new connection that we didn't request!")
+            }
+            self.stats.connecting -= 1
+            self.stats.idle += 1
+            self.stats.availableStreams += maxStreams
+            let connectionInfo = self.connections[index].connected(connection, maxStreams: maxStreams)
+            // TODO: If this is an overflow connection, but we are currently also creating a
+            //       persisted connection, we might want to swap those.
+            let context = self.makeAvailableConnectionContextForConnection(at: index, info: connectionInfo)
+            return (index, context)
+        }
+
+        @inlinable
+        mutating func backoffNextConnectionAttempt(_ connectionID: Connection.ID) -> ConnectionTimer {
+            guard let index = self.connections.firstIndex(where: { $0.id == connectionID }) else {
+                preconditionFailure("We tried to create a new connection that we know nothing about?")
+            }
+
+            self.stats.connecting -= 1
+            self.stats.backingOff += 1
+
+            return self.connections[index].failedToConnect()
+        }
+
+        @usableFromInline
+        enum BackoffDoneAction {
+            case createConnection(ConnectionRequest, TimerCancellationToken?)
+            case cancelTimers(Max2Sequence<TimerCancellationToken>)
```

Comment:
I keep meaning to ask - is `Array` really so non-performant for "collections of 0, 1, or 2 elements" as to make `Max2Sequence` that much worth having?

Reply:
An array of 0 is for free. An array of 2 elements is an allocation. Given how often we run into those situations, I'm happy to save those.